### PR TITLE
Add analyzer support to validate the stateless marshaller shapes

### LINKED ISSUE #72622: MarshallingAttributeAnalyzers throws InvalidOperationException
### Description

NativeMarshallingAttributeAnalyzer and CustomMarshallerAttributeAnalyzer are throwing exceptions when building the runtime repo.

### Reproduction Steps

Set `dotnet_diagnostic.AD0001.severity = error` in `eng/CodeAnalysis.src.globalconfig`, and run `build.cmd`

### Expected behavior

No errors

### Actual behavior

I get lots of errors like the following:

```
CSC : error AD0001: Analyzer 'Microsoft.Interop.Analyzers.CustomMarshallerAttributeAnalyzer' threw an exception of type 'System.InvalidOperationException' with me 
ssage 'Sequence contains no elements'. [C:\Users\svbomer\src\runtime\src\coreclr\System.Private.CoreLib\System.Private.CoreLib.csproj]
```
```
CSC : error AD0001: Analyzer 'Microsoft.Interop.Analyzers.NativeMarshallingAttributeAnalyzer' threw an exception of type 'System.InvalidOperationException' with m 
essage 'Sequence contains no elements'. [C:\Users\svbomer\src\runtime\src\coreclr\System.Private.CoreLib\System.Private.CoreLib.csproj]
```
```
CSC : error AD0001: Analyzer 'Microsoft.Interop.Analyzers.CustomMarshallerAttributeAnalyzer' threw an exception of type 'System.InvalidCastException' with message 'Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Symbols.PublicModel.ArrayTypeSymbol' to type 'Microsoft.CodeAnalysis.INamedTypeSymbol'.'. [C:\Users\svbomer\src\runtime\src\coreclr\nativeaot\System.Private.CoreLib\src\System.Private.CoreLib.csproj]
```

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

@jkoritzinsky

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Support for validating the stateful shapes and for providing the code-fix to add the missing members is left for a future PR to try to keep the size on this one down at least a little bit.

Fixes https://github.com/dotnet/runtime/issues/72622#issuecomment-1191914213 (verified by turning on AD0001 diagnostic locally and making sure that none of the libraries build emits it)

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx [hunk 1]
```diff
@@ -306,23 +318,47 @@
   <data name="ValueInRequiresOneParameterConstructorMessage" xml:space="preserve">
     <value>The type '{0}' specifies that it supports 'In' marshalling of '{1}' but does not provide a one-parameter constructor that takes a '{1}' as a parameter</value>
   </data>
+  <data name="StatelessValueInRequiresConvertToUnmanagedDescription" xml:space="preserve">
+    <value>A 'Value'-kind native type must provide a one-parameter constructor taking the managed type as a parameter</value>
+  </data>
+  <data name="StatelessValueInRequiresConvertToUnmanagedMessage" xml:space="preserve">
+    <value>The type '{0}' specifies that it supports the '{1}' marshal mode of '{2}' but does not provide a one-parameter 'ConvertToUnmanaged' that takes a '{2}' as a parameter and returns a value of an 'unmanaged' type.</value>
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx:325]:
```suggestion
    <value>The type '{0}' specifies that it supports the '{1}' marshal mode for '{2}' but does not provide a one-parameter 'ConvertToUnmanaged' that takes a '{2}' as a parameter and returns a value of an 'unmanaged' type.</value>
```

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx [hunk 2]
```diff
@@ -306,23 +318,47 @@
...
+  </data>
   <data name="LinearCollectionInRequiresCollectionMethodsDescription" xml:space="preserve">
-    <value>A 'LinearCollection'-kind native type that supports marshalling in the 'In' direction must provide a 'GetManagedValuesSource' that returns a 'ReadOnlySpan&lt;&gt;' and a 'GetNativeValuesDestination' method that returns a 'Span&lt;byte&gt;'.</value>
+    <value>A contiguous collection marshaller that supports marshalling from managed to unmanaged must provide a 'GetManagedValuesSource' that returns a 'ReadOnlySpan&lt;&gt;' and a 'GetNativeValuesDestination' method that returns a 'Span&lt;&gt;'.</value>
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx:328]:
```suggestion
    <value>A contiguous collection marshaller that supports marshalling from managed to unmanaged must provide a 'GetManagedValuesSource' that returns a 'ReadOnlySpan&lt;&gt;' and a 'GetUnmanagedValuesDestination' method that returns a 'Span&lt;&gt;'.</value>
```

Comment (elinor-fung) [reply src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx:328]: Throughout for NativeValues -> UnmanagedValues

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx [hunk 3]
```diff
@@ -306,23 +318,47 @@
...
   </data>
   <data name="LinearCollectionOutRequiresCollectionMethodsMessage" xml:space="preserve">
-    <value>The type '{0}' specifies that it supports marshalling in the 'Out' direction, but it does not provide a 'GetManagedValuesDestination' that takes an 'int' and returns a 'Span&lt;&gt;' and a 'GetNativeValuesSource' method that takes an 'int' and rreturns a 'ReadOnlySpan&lt;byte&gt;'</value>
+    <value>The type '{0}' specifies that it supports the '{1}' marshal mode, but it does not provide a 'GetManagedValuesDestination' that takes an 'int' and returns a 'Span&lt;&gt;' and a 'GetNativeValuesSource' method that takes an 'int' and returns a 'ReadOnlySpan&lt;&gt;'</value>
+  </data>
+  <data name="StatelessLinearCollectionOutRequiresCollectionMethodsDescription" xml:space="preserve">
+    <value>A contiguous collection marshaller that supports marshalling from unmanaged to managed must provide a 'GetManagedValuesDestination' that takes the unmanaged value and  an 'int' and returns a 'Span&lt;&gt;' and a 'GetNativeValuesSource' method that takes the unmanaged value and an 'int' and returns a 'ReadOnlySpan&lt;&gt;'.</value>
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx:358]:
```suggestion
    <value>A contiguous collection marshaller that supports marshalling from unmanaged to managed must provide a 'GetManagedValuesDestination' that takes the managed value and returns a 'Span&lt;&gt;' and a 'GetUnmanagedValuesSource' method that takes the unmanaged value and an 'int' and returns a 'ReadOnlySpan&lt;&gt;'.</value>
```

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx [hunk 4]
```diff
@@ -306,23 +318,47 @@
...
+  </data>
+  <data name="StatelessLinearCollectionOutRequiresCollectionMethodsMessage" xml:space="preserve">
+    <value>The type '{0}' specifies that it supports the '{1}' marshal mode, but it does not provide a 'GetManagedValuesDestination' that takes '{2}' and an 'int' and returns a 'Span&lt;&gt;' and a 'GetNativeValuesSource' method that takes the unmanaged value and an 'int' and returns a 'ReadOnlySpan&lt;&gt;'</value>
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx:361]:
```suggestion
    <value>The type '{0}' specifies that it supports the '{1}' marshal mode, but it does not provide a 'GetManagedValuesDestination' that takes '{2}' and returns a 'Span&lt;&gt;' and a 'GetUnmanagedValuesSource' method that takes the unmanaged value and an 'int' and returns a 'ReadOnlySpan&lt;&gt;'</value>
```

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypeSymbolExtensions.cs [hunk 5]
```diff
@@ -277,30 +277,24 @@ public static INamedTypeSymbol ResolveUnboundConstructedTypeToConstructedType(th
...
+        }
 
-            // Get the type arguments and their nullable annotations from the instantiated generic type, including all levels of nesting.
-            // Examples:
-            // Foo<A> -> [A]
-            // Foo.Bar<A> -> [A]
-            // Foo<A?>.Bar<B> -> [A?, B]
-            // Foo<A, B, C?>.Bar<D?, E> -> [A, B, C?, D?, E]
-            static (ImmutableArray<ITypeSymbol>, ImmutableArray<NullableAnnotation>) GetArgumentsToSubstitute(INamedTypeSymbol instantiatedGeneric)
+        public static (ImmutableArray<ITypeSymbol> TypeArguments, ImmutableArray<NullableAnnotation> TypeArgumentNullableAnnotations) GetAllTypeArgumentsIncludingInContainingTypes(this INamedTypeSymbol genericType)
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypeSymbolExtensions.cs:282]: Please comment and fully document what this method is doing and how. We have two loops and populating in a way that would seem order is important.

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeAnalyzer.cs [hunk 6]
```diff
@@ -206,23 +317,139 @@ public static class MissingMemberNames
                 isEnabledByDefault: true,
                 description: GetResourceString(nameof(SR.ManagedTypeMustBeNonNullDescription)));
 
+        // We are intentionally using the same diagnostic ids as the parent type.
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeAnalyzer.cs:320]:
```suggestion
        // We are intentionally using the same diagnostic IDs as the parent type.
```

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx [hunk 7]
```diff
@@ -306,23 +318,47 @@
   <data name="ValueInRequiresOneParameterConstructorMessage" xml:space="preserve">
     <value>The type '{0}' specifies that it supports 'In' marshalling of '{1}' but does not provide a one-parameter constructor that takes a '{1}' as a parameter</value>
   </data>
+  <data name="StatelessValueInRequiresConvertToUnmanagedDescription" xml:space="preserve">
+    <value>A stateless value marshaller that supports marshalling from managed to unmanaged must provide a one-parameter 'ConvertToUnmanaged' method that takes the managed value as the first parameter and returns a value of the 'unmanaged' type.</value>
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx:322]:
```suggestion
    <value>A stateless value marshaller that supports marshalling from managed to unmanaged must provide a one-parameter 'ConvertToUnmanaged' method that takes the managed value as the parameter and returns a value of the 'unmanaged' type.</value>
```

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx [hunk 8]
```diff
@@ -375,6 +411,42 @@
...
+    <value>The first parameter of '{0}' must be the same type as the return type of '{1}'</value>
+  </data>
+  <data name="FirstParametersMustMatchDescription" xml:space="preserve">
+    <value>The first parameters of the two methods must be the same type.</value>
+  </data>
+  <data name="FirstParametersMustMatchMessage" xml:space="preserve">
+    <value>The first parameter of '{0}' and '{1}' must be the same type</value>
+  </data>
+  <data name="ElementTypesOfReturnTypesMustMatchDescription" xml:space="preserve">
+    <value>The element type of the span returned by the first method must be the same type as the element type of the span returned by the second type.</value>
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx:445]:
```suggestion
    <value>The element type of the span returned by the first method must be the same type as the element type of the span returned by the second method.</value>
```

**Review by elinor-fung — APPROVED** [review]

**Pull request merged**